eZ/Publish/Core/FieldType/Image/IO/Legacy.php

```diff
@@ -149,7 +150,13 @@ public function loadBinaryFile($binaryFileId)
     public function loadBinaryFileByUri($binaryFileUri)
     {
         try {
-            return $this->publishedIOService->loadBinaryFileByUri($binaryFileUri);
+            $image = $this->publishedIOService->loadBinaryFileByUri($binaryFileUri);
+
+            if ($image instanceof MissingBinaryFile) {
+                $image = $this->draftIOService->loadBinaryFileByUri($binaryFileUri);
+            }
+
+            return $image;
         } catch (InvalidArgumentException $prefixException) {
             // InvalidArgumentException means that the prefix didn't match, NotFound can pass through
             try {
```

Comment:
my concern here was that theoretically someone can inject his own `publishedIOService` which [according to](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/IO/IOServiceInterface.php#L85) `loadBinaryFileByUri` docblock in `IOServiceInterface` throws `InvalidArgumentException`. In that case when custom `publishedIOService` throws it (instead of returning MissingBinaryFile) we need to catch it and call `draftIOService->loadBinaryFileByUri` - removing  `catch` would be a BC break IMO but question is do we care about such cases?

But you are obviously right with calling it two times. So I see two options here:
1. Don't bother catching and just do:
```
    public function loadBinaryFileByUri($binaryFileUri)
    {
        $image = $this->publishedIOService->loadBinaryFileByUri($binaryFileUri);

        if ($image instanceof MissingBinaryFile) {
            $image = $this->draftIOService->loadBinaryFileByUri($binaryFileUri);
        }

        return $image;
    }
```
2. Stick to possibility `InvalidArgumentException` is thrown and do:
```
    public function loadBinaryFileByUri($binaryFileUri)
    {
        try {
            $image = $this->publishedIOService->loadBinaryFileByUri($binaryFileUri);

            if ($image instanceof MissingBinaryFile) {
                throw new InvalidArgumentException();
            }

            return $image;
        } catch (InvalidArgumentException $prefixException) {
            // InvalidArgumentException means that the prefix didn't match, NotFound can pass through
            try {
                return $this->draftIOService->loadBinaryFileByUri($binaryFileUri);
            } catch (InvalidArgumentException $e) {
                throw $prefixException;
            }
        }
    }
```

What do you think?